File: EdDyA II/HashTables/Set.py
class Set:
  def __init__(self):
    self.si = 1000
    self.hasht = [[]]
    for i in range(self.si):
      self.hasht.append([])
  def getVal(self, dato):
    if isinstance(dato, int):
      return dato % self.si
    s = 0
    for i in range(len(dato)):
      s += ord(dato[i])*(26**i)
      s %= self.si
    return s
  def insert(self, v):
     pos = self.getVal(v)
     if not v in self.hasht[pos]:
       self.hasht[pos].append(v)
  def remove(self, v):
    pos = self.getVal(v)
    if v in self.hasht[pos]:
      self.hasht[pos].remove(v)
  def contains(self, v):
    pos = self.getVal(v)
    return v in self.hasht[pos]
  def __str__(self):
    s = ""
    for i in self.hasht:
      if i:
        for j in i:
          s += "(" + str(j) + ")"
    return s + "\n"
s = Set()
print(s.contains("hola"))
s.insert("hola")
s.insert(10)
print(s.contains(100))
print(s.contains("hola"))
print(s)
s.remove("adios")
s.remove("hola")
print(s)
s.remove(10)
print(s)

File: EdDyAII/HashTables/Set.cpp
#include <iostream>
#include <vector>
#include <list>
#include <math.h>
#include <algorithm>
using namespace std;
template<class value>
class Set{
  public: 
    int si;
    vector<list<value>> hash;
    Set(){
      si = 1000;
      hash.resize(si);
    }
    int getVal(string s){
      long long a = 0;
      for(int i = 0; i < s.size(); i++)
	a += s[i]*pow(26, i), a %= si;
      return (int)a;
    }
    int getVal(long long a){
      return a % si;
    }
    void insert(value v){
      int pos = getVal(v);
      if(find(hash[pos].begin(), hash[pos].end(), v) == hash[pos].end())
	hash[pos].push_back(v);
    }
    void remove(value v){
      int pos = getVal(v);
      hash[pos].remove(v);
    }
    bool contains(value v){
      int pos = getVal(v);
      return (find(hash[pos].begin(), hash[pos].end(), v) != hash[pos].end());
    }
    friend ostream &operator << (ostream &output, const Set &s){
      for(int i = 0; i < s.si; i++)
	if(s.hash[i].size())
	  for(auto j: s.hash[i])
	    output << "\t(" << j << ")" << endl;
      return output;
    }
    ~Set(){}
};
int main(){
  Set<int> s;
  cout << s.contains(1010) << endl;
  s.insert(1010);
  cout << s.contains(1010) << endl;
  s.insert(1010);
  cout << s.contains(1010) << endl;
  cout << s << endl;
  s.remove(1010); 
  cout << s << endl;
  s.insert(1);
  s.insert(2);
  cout << s << endl;
  s.remove(1011);
  cout << s << endl;
  /*
  Set<string> s;
  cout << s.contains("hola") << endl;
  s.insert("hola");
  cout << s.contains("hola") << endl;
  s.insert("hola");
  cout << s.contains("hola") << endl;
  cout << s << endl;
  s.remove("hola"); 
  cout << s << endl;
  s.insert("asi");
  s.insert("es");
  cout << s << endl;
  s.remove("asiera");
  cout << s << endl;
  */
  return 0;
}
    
